osquery/tables/system/darwin/battery.mm

```diff
@@ -0,0 +1,192 @@
+
+  CFRelease(list);
+  CFRelease(info);
+
+  return (__bridge_transfer NSDictionary*)battery;
+}
+
+NSDictionary* getIopmpsBatteryInfo() {
+  CFMutableDictionaryRef matching, properties = NULL;
+  io_registry_entry_t entry = 0;
+  matching = IOServiceNameMatching("AppleSmartBattery");
+  entry = IOServiceGetMatchingService(kIOMasterPortDefault, matching);
+  IORegistryEntryCreateCFProperties(entry, &properties, NULL, 0);
+  return (__bridge_transfer NSDictionary*)properties;
+}
+
+QueryData genBatteryInfo(QueryContext& context) {
+  QueryData results;
+  Row r;
+
+  NSDictionary* batteryInfo = getIopmBatteryInfo();
+  NSDictionary* advancedBatteryInfo = getIopmpsBatteryInfo();
+
+  // Don't return any rows if we don't have battery data.
+  if (batteryInfo == NULL && advancedBatteryInfo == NULL) {
+    return results;
+  }
+
+  if ([advancedBatteryInfo objectForKey:@"Manufacturer"]) {
+    r["manufacturer"] =
+        TEXT([[advancedBatteryInfo objectForKey:@"Manufacturer"] UTF8String]);
+  }
+
+  if ([advancedBatteryInfo objectForKey:@kIOPMPSManufactureDateKey]) {
+    // Date is published in a bitfield per the Smart Battery Data spec rev 1.1
+    // in section 5.1.26 Bits 0...4 => day (value 1-31; 5 bits) Bits 5...8 =>
+    // month (value 1-12; 4 bits) Bits 9...15 => years since 1980 (value 0-127;
+    // 7 bits)
+    int dateMask = [
+        [advancedBatteryInfo objectForKey:@kIOPMPSManufactureDateKey] intValue];
+    int day = dateMask & 31;
+    int month = (dateMask >> 5) & 15;
+    int year = (dateMask >> 9) + 1980;
+
+    NSCalendar* calendar = [[NSCalendar alloc]
+        initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
+    calendar.timeZone = [NSTimeZone timeZoneWithAbbreviation:@"UTC"];
+    NSDateComponents* components = [[NSDateComponents alloc] init];
+    [components setDay:day];
+    [components setMonth:month];
+    [components setYear:year];
+    NSDate* date = [calendar dateFromComponents:components];
+
+    r["manufacture_date"] = INTEGER([date timeIntervalSince1970]);
+  }
+
+  if ([advancedBatteryInfo objectForKey:@kIOPMDeviceNameKey]) {
+    r["model"] = TEXT(
+        [[advancedBatteryInfo objectForKey:@kIOPMDeviceNameKey] UTF8String]);
+  }
+
+  if ([batteryInfo objectForKey:@kIOPSHardwareSerialNumberKey]) {
+    r["serial_number"] = TEXT(
+        [[batteryInfo objectForKey:@kIOPSHardwareSerialNumberKey] UTF8String]);
+  }
+
+  if ([advancedBatteryInfo objectForKey:@"CycleCount"]) {
+    r["cycle_count"] =
+        INTEGER([[advancedBatteryInfo objectForKey:@"CycleCount"] intValue]);
+  }
+
+  if ([batteryInfo objectForKey:@kIOPSBatteryHealthKey]) {
+    r["health"] =
+        TEXT([[batteryInfo objectForKey:@kIOPSBatteryHealthKey] UTF8String]);
+  }
+
+  if ([batteryInfo objectForKey:@kIOPSBatteryHealthConditionKey]) {
+    r["condition"] = TEXT([
+        [batteryInfo objectForKey:@kIOPSBatteryHealthConditionKey] UTF8String]);
+  } else {
+    r["condition"] = TEXT("Normal");
```

Comment:
I assume we hit this if that function doesn't return anything. Is there a reason you want to return normal instead of just nothing?

Reply:
Yes, apple's own tools return "Normal" when this data is unavailable. Since it is only populated when there is an issue with the condition, unlike `health` which is populated always.